src/board.rs

```diff
@@ -87,40 +87,104 @@ impl Board {
+        let direction = match player {
+            Player::White => -1,
+            Player::Black => 1,
+        };
+        let advance_moves = match player {
+            Player::White => Move::get_pawn_advance_moves_white(),
+            Player::Black => Move::get_pawn_advance_moves_black(),
+        };
+
+        for &m in advance_moves {
+            let new_position = Position {
+                x: start.x + (direction * m.dx) as usize,
+                y: start.y + m.dy as usize,
+            };
+            if Self::is_in_bounds(new_position).is_ok() && self.get_piece(new_position).is_none() {
+                self.moves.insert((start, new_position));
+                if ((player == Player::White && start.x == 6)
+                    || (player == Player::Black && start.x == 1))
+                    && self
+                        .get_piece(Position {
+                            x: start.x + (2 * direction * m.dx) as usize,
+                            y: start.y,
+                        })
+                        .is_none()
+                {
+                    self.moves.insert((
+                        start,
+                        Position {
+                            x: start.x + (2 * direction * m.dx) as usize,
+                            y: start.y,
+                        },
+                    ));
                 }
-                return;
             }
-            self.moves.insert((start, position));
-            // if the piece cannot snipe (move multiple moves in one direction),
-            // return after the first move
-            if !self.get_piece(start).unwrap().can_snipe() {
-                return;
+        }
+    }
+
+    fn add_pawn_capture_moves(&mut self, start: Position, player: Player) {
+        let direction = match player {
+            Player::White => -1,
+            Player::Black => 1,
+        };
+        let capture_moves = match player {
+            Player::White => Move::get_pawn_capture_moves_white(),
+            Player::Black => Move::get_pawn_capture_moves_black(),
+        };
+
+        for &m in capture_moves {
+            let new_position = Position {
+                x: start.x + (direction * m.dx) as usize,
+                y: start.y + m.dy as usize,
+            };
+            if Self::is_in_bounds(new_position).is_ok() {
+                if let Some(other_piece) = self.get_piece(new_position) {
+                    if other_piece.get_player() != player {
+                        self.moves.insert((start, new_position));
+                    }
+                }
+            }
+        }
+    }
+
+    fn add_moves_in_direction(&mut self, start: Position, piece: Piece) {
+        match piece {
+            Piece::Pawn(player) => {
+                self.add_pawn_advance_moves(start, player);
+                self.add_pawn_capture_moves(start, player);
+            }
+            _ => {
+                for &m in piece.get_moves() {
```

Comment:
Nice, I like `&m` here better

Reply:
Yeah but don't get confused. It's pattern matching on a reference, `m` inside the for loop is actually `*m`